### ISSUE TITLE
Injecting and extracting traceparent for distributed tracing

### ISSUE THREAD

**Issue by tyhill3** [open-telemetry/opentelemetry-erlang#694]
Hello. I currently have two services in Kubernetes. This first service is the one I plan to inject the traceparent on each HTTP request with this plug:

```
defmodule MyFirstAppWeb.Plug.OpenTelemetrySpanInjector do
  @moduledoc false

  @behaviour Plug

  import Plug.Conn

  require OpenTelemetry.Tracer, as: Tracer

  def init(opts), do: opts

  def call(conn, _opts) do
    Tracer.with_span conn.request_path do
      merged_headers = :otel_propagator_text_map.inject(conn.req_headers)

      trace_parent_id =
        case Enum.find_value(merged_headers, fn
               {"traceparent", value} -> value
               _ -> nil
             end) do
          nil -> nil
          trace_parent_id -> trace_parent_id
        end

      conn = put_req_header(conn, "traceparent", trace_parent_id)
      conn
    end
  end
end
```

When I do an `IO.inspect(conn.req_headers)` before I send the conn in the above module, I definitely see the `traceparent` header, so I believe it's being sent.

The second service is the service I want to extract the traceparent so I can see distributed tracing in Tempo. I don't know if I'm doing this part right, as I am not getting distributed tracing in Tempo, and each service only has their own spans. If I do `IO.inspect(conn.req_headers)` on the incoming request from the first service here, I do not see the `traceparent` header. Here is the plug for this one:

```
defmodule MySecondAppWeb.Plug.OpenTelemetrySpanInjector do
  @moduledoc false

  @behaviour Plug

  import Plug.Conn

  require OpenTelemetry.Tracer

  def init(opts), do: opts

  def call(conn, _opts) do
    :otel_propagator_text_map.extract(conn.req_headers)

    span_ctx = OpenTelemetry.Tracer.start_span(<<"span-name">>)
    ctx = OpenTelemetry.Ctx.get_current()

    OpenTelemetry.Ctx.attach(ctx)
    OpenTelemetry.Tracer.set_current_span(ctx)
    OpenTelemetry.Tracer.end_span(span_ctx)

    conn
  end
end
```

Any help on this?

**Comment by tsloughter** [comment]
You should be able to just do:

```
    :otel_propagator_text_map.extract(conn.req_headers)

    OpenTelemetry.Tracer.start_span(<<"span-name">>)
   
    OpenTelemetry.Tracer.end_span()
```

`extract` will use the context in the pdict, as will the macros `start_span` and `end_span`.

**Comment by tsloughter** [comment]
Oh, I see the original issue, since `end_span` is a macro that works off the context of the pdict you don't pass a `span_ctx`, the only arg it takes is an end timestamp if you need to set that manually. So your call to `end_span(span_ctx)` isn't actually ending the span.

**Comment by tsloughter** [comment]
Also note that outside of a Plug you'll want to use `Tracer.with_span` which takes a `fun` and handles context starting and ending for you. In a Plug you'll need to either use `start_span` without an explicit context as you are doing or with an explicit `Ctx` and set it in a `conn` to be read in later Plugs.

**Comment by tyhill3** [comment]
So my second service now looks:

```
defmodule MySecondAppWeb.Plug.OpenTelemetrySpanInjector do
  @moduledoc false

  @behaviour Plug

  import Plug.Conn

  require OpenTelemetry.Tracer

  def init(opts), do: opts

  def call(conn, _opts) do
    :otel_propagator_text_map.extract(conn.req_headers)

    OpenTelemetry.Tracer.start_span(<<"span-name">>)

    OpenTelemetry.Tracer.end_span()

    conn
  end
end
```

And now I'm wondering if the first service that does the `inject` is not correct?

```
defmodule MyFirstAppWeb.Plug.OpenTelemetrySpanInjector do
  @moduledoc false

  @behaviour Plug

  import Plug.Conn

  require OpenTelemetry.Tracer, as: Tracer

  def init(opts), do: opts

  def call(conn, _opts) do
    Tracer.with_span conn.request_path do
      :otel_propagator_text_map.inject([])
      conn
    end
  end
end
```

Do I need the traceparent in the headers like in my original post?

```
       trace_parent_id =
         case Enum.find_value(merged_headers, fn
                {"traceparent", value} -> value
               _ -> nil
              end) do
           nil -> nil
           trace_parent_id -> trace_parent_id
         end

       conn = put_req_header(conn, "traceparent", trace_parent_id)
       IO.inspect(conn.req_headers)
      ```

**Comment by tyhill3** [comment]
Or since it's in a Plug, should I be using the `start_span` and `end_span` like you mentioned above? And since this is the "initiating" call, would the context need to be set in this Plug? If so, how do I go about that?

**Comment by tsloughter** [comment]
That Plug to inject won't work because `inject` returns the list of headers you then have to pass to a request to send them to another service.

**Comment by tyhill3** [comment]
But adding the `traceparent` header in the plug doesn't work either when I pass those headers to a request to send them to another service. 

```
trace_parent_id =
         case Enum.find_value(merged_headers, fn
                {"traceparent", value} -> value
               _ -> nil
              end) do
           nil -> nil
           trace_parent_id -> trace_parent_id
         end

       conn = put_req_header(conn, "traceparent", trace_parent_id)
       ```
       
       I mean I don't see the `traceparent` header in the other service.

**Comment by tyhill3** [comment]
So I ended up finding out that the `:otel_propagator_text_map.inject(headers)` works just fine for injecting `traceparent` into the conn.req_headers. I was originally trying to do this injection in a Plug, but I found out that we use Finch, and we set our own headers there before sending a request to the other service.

So doing the injection in our custom Finch module worked, and I can now see the `traceparent` header in the other service via a Plug that does the extraction with the existing context.

Thanks for much for your help!
